### PR TITLE
Check and apply max deduction for handgrading annotations

### LINKED ISSUE #502: Local score update when applying annotations past the max deduction is incorrect
If an annotation that has a max deduction set is applied multiple times, the score on the page keeps going down even though the value on the server is correct.

### PR DESCRIPTION
**Pull request by broad-well**
This PR fixes #502. When an annotation is applied or deleted in the Handgrading interface, we compute the effective change to the score based on its max deduction and total number of applications so far.

Tests are included for both applying and deleting annotations across the max deduction boundary. Additional tests are included for cases where the max deduction of an annotation is not a multiple of its unit deduction, such as ("-3/-7 max"). This PR's implementation changes makes those tests pass.

## Screenshots

**After**

Screenshot 2024-04-20 at 12-34-51 Autograder io

### CODE REVIEW

File: tests/test_components/test_project_view/test_handgrading/test_handgrading.ts
Comment (james-perretta) [inline tests/test_components/test_project_view/test_handgrading/test_handgrading.ts]: I tend to avoid `$forceUpdate` to make sure that the page is re-rendered appropriately. Does it work with `expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);` instead?

Comment (Normanwqn) [reply tests/test_components/test_project_view/test_handgrading/test_handgrading.ts]: I will test with that. Is there a reason that `$forceUpdate` is used here?
```
 test('Handgraders allowed to leave custom comments', async () => {
        wrapper.vm.d_handgrading_result!.handgrading_rubric.handgraders_can_leave_comments = true;
        data_ut.set_global_user_roles(data_ut.make_user_roles({is_handgrader: true}));
        wrapper.vm.$forceUpdate();
        await wrapper.vm.$nextTick();
        expect(wrapper.find('#new-comment').exists()).toBe(true);

        expect(wrapper.findAll('.comment').at(0).find('.delete').exists()).toBe(true);
        expect(wrapper.findAll('.comment').at(3).find('.delete').exists()).toBe(true);
        // Can always delete applied annotations
        expect(wrapper.findAll('.comment').at(1).find('.delete').exists()).toBe(true);
        expect(wrapper.findAll('.comment').at(2).find('.delete').exists()).toBe(true);
    });
  ```

**Comment by Normanwqn** [comment]
@james-perretta
We changed `$forceUpdate` to `expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);`, and all unit tests are still passed

We also changed a previously written test case `Handgraders allowed to leave custom comments` which uses `$forceUpdate`, to use ` `expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);` instead.

**Review by james-perretta — DISMISSED** [review]

**Review by james-perretta — DISMISSED** [review]

**Comment by james-perretta** [comment]
@broad-well Did github do that automatically (dismissing the stale review)? For some reason the "merge automatically" check isn't going through on my end, and it still says review required after I approved the PR. I'll check back later to see if it resolves itself, otherwise I'll have you try re-opening the PR.

**Comment by broad-well** [comment]
> @broad-well Did github do that automatically (dismissing the stale review)? For some reason the "merge automatically" check isn't going through on my end, and it still says review required after I approved the PR. I'll check back later to see if it resolves itself, otherwise I'll have you try re-opening the PR.

Yes, GitHub automatically dismissed your review. It does this because (1) I branched off to create `handgrading` from `develop` in my fork, (2) I started a PR from my `handgrading` branch to the origin's `develop` branch, and (3) a PR from my `develop` branch to the origin's `develop` branch got merged. [Other users have had the same issue](https://github.com/orgs/community/discussions/58535#discussioncomment-6495938).

We can either duplicate this pull request or have you temporarily turn off "Dismiss stale pull request approvals when new commits are pushed".

**Comment by james-perretta** [comment]
Got it, go ahead and make a fresh PR please

**Comment by Normanwqn** [comment]
@james-perretta  We have submitted another PR <a href="https://github.com/eecs-autograder/ag-website-vue/pull/514"> here </a>. We wold appreciate it if you may take a look. Thank you so much!

**Pull request closed without merging**
